Source/Services/ReverseCalls/WrappedAsyncStreamWriter.cs

```diff
@@ -68,37 +77,81 @@ public WriteOptions WriteOptions
         /// <returns>A task that will complete when the message is sent.</returns>
         public async Task WriteAsync(TServerMessage message)
         {
+            _logger.WritingMessage(_requestId, typeof(TServerMessage));
             _cancellationToken.ThrowIfCancellationRequested();
+
             if (!_writeLock.Wait(0))
             {
                 var stopwatch = Stopwatch.StartNew();
+
                 _metrics.IncrementPendingStreamWrites();
+                _logger.WritingMessageBlockedByAnotherWrite(_requestId, typeof(TServerMessage));
+
                 await _writeLock.WaitAsync(_cancellationToken).ConfigureAwait(false);
-                _metrics.DecrementPendingStreamWrites();
+
                 stopwatch.Stop();
+                _metrics.DecrementPendingStreamWrites();
                 _metrics.AddToTotalStreamWriteWaitTime(stopwatch.Elapsed);
+                _logger.WritingMessageUnblockedAfter(_requestId, typeof(TServerMessage), stopwatch.Elapsed);
             }
 
-            await WriteRecordMetricsAndReleaseLock(message, false);
+            try
+            {
```

Comment:
This should perhaps be moved right under the acquiring of the lock

Reply:
Will fix and make another PR